builtin/builtin_loader.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+package builtin
+
+import (
+	"fmt"
+
+	"github.com/veraison/services/log"
+	"github.com/veraison/services/plugin"
+	"go.uber.org/zap"
+)
+
+var defaultBuiltinLoader *BuiltinLoader
+
+type BuiltinLoader struct {
+	logger *zap.SugaredLogger
+
+	loadedByName      map[string]plugin.IPluggable
+	loadedByMediaType map[string]plugin.IPluggable
+
+	registeredPluginTypes map[string]string
+}
+
+func NewBuiltinLoader(logger *zap.SugaredLogger) *BuiltinLoader {
+	return &BuiltinLoader{logger: logger}
+}
+
+func CreateBultinLoader(
+	cfg map[string]interface{},
+	logger *zap.SugaredLogger,
+) (*BuiltinLoader, error) {
+	loader := NewBuiltinLoader(logger)
+	err := loader.Init(cfg)
+	return loader, err
+}
+
+func (o *BuiltinLoader) Init(m map[string]interface{}) error {
+	o.loadedByName = map[string]plugin.IPluggable{}
+
+	o.loadedByMediaType = make(map[string]plugin.IPluggable)
+	o.registeredPluginTypes = make(map[string]string)
+	return nil
+}
+
+func (o *BuiltinLoader) GetRegisteredMediaTypes() []string {
+	var mediaTypes []string // nolint:prealloc
+
+	for mt := range o.loadedByMediaType {
+		mediaTypes = append(mediaTypes, mt)
+	}
+
+	return mediaTypes
+}
+
+func DiscoverBuiltin[I plugin.IPluggable]() error {
+	return DiscoverBuiltinUsing[I](defaultBuiltinLoader)
+}
+
+func DiscoverBuiltinUsing[I plugin.IPluggable](loader *BuiltinLoader) error {
+	for _, p := range plugins {
+		name := p.GetName()
+		if _, ok := loader.loadedByName[name]; ok {
+			loader.logger.Panicw("duplicate plugin name", "name", name)
+		}
+
+		loader.logger.Debugw("found scheme", "name", name)
+		loader.loadedByName[name] = p
+
+		for _, mt := range p.GetSupportedMediaTypes() {
+			if existing, ok := loader.loadedByMediaType[mt]; ok {
+				loader.logger.Panicw("media type handled by two plugins",
+					"media-type", mt,
+					"plugin1", existing.GetName(), "plugin2", name)
+			}
+
+			loader.logger.Debugw("media type handled", "media-type", mt, "name", name)
+			loader.loadedByMediaType[mt] = p
+		}
+	}
+
+	return nil
+}
+
+func GetBuiltinHandleByMediaType[I plugin.IPluggable](mediaType string) (I, error) {
+	return GetBuiltinHandleByMediaTypeUsing[I](defaultBuiltinLoader, mediaType)
+}
+
+func GetBuiltinHandleByMediaTypeUsing[I plugin.IPluggable](
```

Comment:
func GetBuiltinHandleByMediaTypeUsing[I plugin.IPluggable]( 
Did you mean to reword it with taking into consideration Using ??

Reply:
The "Using" is gontinued by the first paramer which lis the loader.  (see the call in line 85). I could say "UsingLoader", but that would just make an already long function longer without adding additional information.